### PR TITLE
Change AdminMerge to not require right-hand range descriptor.

### PR DESCRIPTION
**Pull request by spencerkimball**
This is much less awkward and more likely to work properly in practice.

### CODE REVIEW

**Comment by BramGruneir** [comment]
LGTM

File: proto/api.proto [hunk 1]
```diff
@@ -408,16 +408,16 @@ message AdminSplitResponse {
...
-// the subsuming range will now encompass all keys from its original
-// start_key to the end_key of the subsumed_range.
+// merge is performed by calling AdminMerge on the left-hand range of
+// two consecutive ranges (i.e. the range which contains keys which
+// sort first). This range will be the subsuming range and the right
+// hand range will be subsumed. After the merge operation, the
+// subsumed_range will no longer exist and the subsuming range will
+// now encompass all keys from its original start_key to the end_key
+// of the subsumed_range. If AdminMerge is called on the final range
+// in the key space, it is a noop.
```
Comment (tschottdorf) [inline proto/api.proto]: Another good place to mention that the two ranges need to be colocated on their replica sets.

**Comment by tschottdorf** [comment]
LGTM

File: proto/api.proto [hunk 2]
```diff
@@ -408,16 +408,16 @@ message AdminSplitResponse {
...
-// be consecutive in the key space, such that the end_key of the
-// subsuming range must match the start_key of the range being subsumed.
-// After the merge operation, the subsumed_range will no longer exist and
-// the subsuming range will now encompass all keys from its original
-// start_key to the end_key of the subsumed_range.
+// merge is performed by calling AdminMerge on the left-hand range of
+// two consecutive ranges (i.e. the range which contains keys which
+// sort first). This range will be the subsuming range and the right
+// hand range will be subsumed. After the merge operation, the
+// subsumed_range will no longer exist and the subsuming range will
```
Comment (mrtracy) [inline proto/api.proto]: s/subsumed_range/subsumed range/

Comment (spencerkimball) [inline proto/api.proto]: Done

File: proto/api.proto (hunk 1 above)
Comment (spencerkimball) [inline proto/api.proto]: Actually, this won't be a requirement at this level. Calling AdminMerge will start an op on the server owning the subsuming range which will start by moving the replicas to the range and then will execute the merge.

**Comment by bdarnell** [comment]
LGTM (I think the test failure will go away if you rebase)

**Comment by spencerkimball** [comment]
@bdarnell I can't seem to figure out this failing test. I'm arbitrarily removing a range from the store which isn't really a valid thing to do under normal operating circumstances, but it serves the purpose of this unittest. However, as you see from the failure, it's complaining about leaking a store. Shouldn't Store.RemoveRange disconnect the range sufficiently to pass the leak tests?

**Comment by bdarnell** [comment]
In the current build (https://circleci.com/gh/cockroachdb/cockroach/1191) there are two failures: the leak and then "close of closed channel". RemoveRange stops the Range, and then when you re-add it the store stops it again. What happens if you simply remove the AddRange call at the end of the test? If you still need it there, you'll need to either make Ranges restartable (by re-creating the closer channel) or figure out another way to get this effect for testing.

**Comment by spencerkimball** [comment]
I started with it removed only. That failed, so I tried adding it back which didn't help. Then removed again just now and re-pushed. I can't remember exactly what the error was, but it'll surface again in a sec.

**Comment by spencerkimball** [comment]
Looks like the problem is a nil pointer reference in multiraft while test is trying to stop.

**Comment by bdarnell** [comment]
At shutdown TxnCoordSender tries to resolve any remaining intents in background goroutines, and this appears to be racing with the rest of the shutdown (it's timing-sensitive - it appears to happen in race mode on circleci, but I checked out your branch locally and can't reproduce it). I'm not sure how it gets to this panic - ProposeRaftCommand will actually recreate the MultiRaft group if necessary (and it looks like we can see this in the logs), so I don't know how we get here without the range in the map.

In any case, if we simply returned an error here instead of a panic things would be fine. The problem is that the async proposal interface doesn't give us a way to return an error. I need to fix this soon anyway as a part of #543 (to handle the automatic retry logic). In the meantime, I think this error will also be fixed when you merge #573 (automatically resolve intents for splits and merges) and rebase on top of it.

**Pull request merged**